### PR TITLE
Speedy cleanup dynamic casting in builtins

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -317,53 +430,50 @@ private[lf] object SBuiltin {
...
-            case _: OffLedger.type =>
-              machine.returnValue = SOptional(Some(SText(cid.coid)))
-          }
-        case _ => crash(s"type mismatch toTextContractId: $args")
+    ): Unit =
+      machine.ledgerMode match {
+        case OffLedger =>
+          machine.returnValue = SOptional(Some(SText(getSContractId(args, 0).coid)))
+        case _ =>
+          machine.returnValue = SValue.SValue.None
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:451]: This is a change in semantic, you now won’t crash if this is used incorrectly in on-ledger mode. I don’t really see an issue with this but was this intentional?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -747,33 +782,24 @@ private[lf] object SBuiltin {
...
-          crash(s"RecUpd on non-record: $v")
+      val record = getSRecord(args, 0)
+      if (record.id != id) {
+        crash(s"type mismatch on record update: expected $id, got record of type ${record.id}")
       }
+      val values2 = record.values.clone.asInstanceOf[util.ArrayList[SValue]]
+      var i = 0
+      while (i < updateFields.length) {
+        values2.set(updateFields(i), args.get(i + 1))
+        i += 1
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:793]:
```suggestion
      for (i <- 0 until updateFields.length) {
        values2.set(updateFields(i), args.get(i + 1))
```

**Review by cocreature — APPROVED** [review]
Thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 3]
```diff
@@ -43,6 +43,139 @@ private[speedy] sealed abstract class SBuiltin(val arity: Int) {
     * Updates the machine state accordingly.
     */
   private[speedy] def execute(args: util.ArrayList[SValue], machine: Machine): Unit
+
+  final protected def getSBool(args: util.ArrayList[SValue], i: Int): Boolean =
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:47]: for another PR.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:451]: It was intentional. i will revert it as it is not really in the general scope of the PR.

Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:451]: fine to keep it as well, just wanted to doublecheck

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:795]: maybe for another PR.

**Pull request merged**